### ISSUE TITLE
[WIKI][RFC] Change video caching defaults for single-board computers?

### ISSUE THREAD

**Issue by markg85** [xbmc/xbmc#22339]
### Describe the bug
KODI's cache settings cannot (easily) be changed by users.

I'll happily refer to https://github.com/xbmc/xbmc/issues/22332 for extra information which was so rudely closed.

KODI's default cache settings are wrong by default. Sure, they work fine for web resources.
But if you start using **local media** then you **will** hit buffering issues with at least a few common kodi places (android tv, windows, arm SBCs). This issue has been going on for many years through many kodi versions. Network throughput is **not** an issue, see #22332 where i proof that.

## Expected Behavior

KODI should expose the [cache settings](https://kodi.wiki/view/HOW-TO:Modify_the_video_cache) in the user interface **without** the need to install a third party addon.

## Actual Behavior

Right now a different way of editing your cache settings is needed. Depending on how you use kodi, this could be insanely complex. On android tv for example, this is nearly impossible to achieve! 

## Possible Fix

Expose [cache settings](https://kodi.wiki/view/HOW-TO:Modify_the_video_cache) in an appropriate menu.
It could be part of `Settings -> Player -> Video`  though that would be inaccurate (cache affects video and audio).
A more appropriate place would be `Settings -> Player -> Cache` (this one doesn't exist yet).

### To Reproduce

Google the many thousands of kodi buffering issues.

### Debuglog

N/A

### Screenshots 
N/A

## Additional context or screenshots (if appropriate)
N/A

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] macOS
 - [x] Windows
 - [ ] Windows UWP

And on many different SBC (Single board computer) with arm as architecture. These seem suspiciously sensitive to this issue.

 - Operating system version/name: N/A
 - Kodi version: All

**Comment by thexai** [comment]
This is not a bug but feature request.

By the way, I use Kodi for local media (NAS) only with default cache settings and have zero issues (Shield, Windows x64 and Xbox clients). NFS and SMB protocols and some files are really big (4K HDR + TrueHD) up to 100 Mbps or more.

**Comment by markg85** [comment]
This is a **real** issue on **many** devices!
This should be a prio 1 blocker bug! Not a feature request.

**Comment by basilgello** [comment]
@markg85 Let's not duplicate issues and get back to the one I re-opened.

**Comment by thexai** [comment]
"Allow users to change cache setting" is a feature 

A bug cannot be something that is not yet implemented.

**Comment by markg85** [comment]
> "Allow users to change cache setting" is a feature
> 
> A bug cannot be something that is not yet implemented.

It's a means to an end.
Fix buffering issues.

Please reopen.

Alternatively you can change the defaults.
https://kodi.wiki/view/HOW-TO:Modify_the_video_cache

The default right now is 0 for the buffermode where i believe 1 is already better.
But you'd have to test this with people who have this issue. (i do)



> @markg85 Let's not duplicate issues and get back to the one I re-opened.

@basilgello I made this one because the other complaint was me not following the template.. sigh. (it wasn't meant as bug report but instead was meant as investigative issue)
Also this is specific to 1 change (cache setting) while the other is more generic. The two would complement each other i think.

**Comment by basilgello** [comment]
@markg85 OK, let's discuss the wiki change proposal here.

So this issue is about buffering of local media? And that one is about network?

**Comment by markg85** [comment]
This issue:
Let kodi use proper buffering. Mode 0 seems to cause buffering issues on **networked protocols** like SMB, NFS and FTP. So the result of this issue is to merely use proper buffering in the majority of cases.

The other issue: #22333
(where i can't reply anymore btw, not even after you re-opened it)
is about figuring out why there is buffering issues in the first place.

Fixing either issue would fix buffering. Fixing the defaults would "mask" it out of existence. Probably not a good way but then again also probably the best achievable.

**Comment by basilgello** [comment]
@markg85 It looks I must figure out how to unlock that issue (btw, wrong number referenced!)

**Comment by markg85** [comment]
The irony is that KODI's own wiki page https://kodi.wiki/view/HOW-TO:Modify_the_video_cache says:
`These are just examples to explain how the feature works. Most users will just want to use example 4.`

Which is:
```
<advancedsettings>
  <cache>
    <buffermode>1</buffermode>
    <memorysize>139460608</memorysize>
    <readfactor>20</readfactor>
  </cache>
</advancedsettings>
```

And is completely different then the **defaults** that kodi uses, those are described in the [table](https://kodi.wiki/view/HOW-TO:Modify_the_video_cache#Cache_settings) on right side. Again confusingly not even on the left with the xml properties, those seem yet again random different values.

**Comment by markg85** [comment]
Here's a suggestion that likely improves the situation for everyone.
cc @thexai 
Just follow your - KODI's - own advice on it's own wiki and _mostly_ pick the defaults of `example 4`
```
<advancedsettings>
  <cache>
    <buffermode>1</buffermode>
    <memorysize>139460608</memorysize>
    <readfactor>20</readfactor>
  </cache>
</advancedsettings>
```

The readfactor does seem a bit high to me though. I'd probably leave that at 4.
Would a patch for that be accepted?

**Comment by basilgello** [comment]
Wiki is outdated in so many places, and not many people are willing to fix it. That's good that you are among those who reads Wiki and can fix it :)

We may write something like:

```
If you experience lags and stutters on commodity hardware (Android TV, Raspberry Pi or othee single-board computers or simply older PCs and HTPCs), you may set…
```

**Comment by basilgello** [comment]
@markg85 to be honest I like the cache setting exposed in GUI + wiki edit better than changing the system-wide defaults. At least, not in v20 nearing its prime-time :)  In v21, probably a better long-term solution would be settings editor or the separate setting like you initially thought of.

**Comment by markg85** [comment]
Oke, this blows my mind...

The codebase and wiki don't match in claimed defaults either.... men what a mess!

```cpp
#define CACHE_BUFFER_MODE_INTERNET 0
#define CACHE_BUFFER_MODE_ALL 1
#define CACHE_BUFFER_MODE_TRUE_INTERNET 2
#define CACHE_BUFFER_MODE_NONE 3
#define CACHE_BUFFER_MODE_NETWORK 4
```

The actual default in code:
```cpp
  m_cacheMemSize = 1024 * 1024 * 20; // 20 MiB
  m_cacheBufferMode = CACHE_BUFFER_MODE_NETWORK; // Default (buffer all network filesystems)
  m_cacheChunkSize = 128 * 1024; // 128 KiB

  // the following setting determines the readRate of a player data
  // as multiply of the default data read rate
  m_cacheReadFactor = 4.0f;
```

Which maps to mode 4 on the wiki.
I'm so confused right now.
Also about why the custom setting i applied - thinking kodi was using mode 0 - even gives a notable change...???

**Comment by basilgello** [comment]
Maybe less capable hardware with slower SD cards or HDD interfaces also needs caching. I still have to look through code what differs modes 0 and 3…

**Comment by markg85** [comment]
In this specific case we're talking about the [Mecool KM7](https://www.mecool.com/products/2021-mecool-km7-android-tv-box-android-11-prime-video-youtube-supported-4gb-ram-64gb-rom) with a rather powerful [Amlogic S905y4](https://androidpctv.com/amlogic-s905y4-comparative/)

There is no SD card or HDD interface in this mix. It has eMMC.
Now it's till android with selectively optimized features (like hardware video decoding is awesome).
I was wrong about network throughput though. I thought it had gigabit but it has 10/100 mbit. I do get high up in those speeds (like ~80 mbit/sec) when testing it a file transfer in another app.

We're still talking about SMB here in kodi, not local on-device storage.

**Comment by fritsch** [comment]
Take care: Kodi has two caches. One aiming to optimize http WebDAV Sharing and others. One inside Videoplayer which you most of the time don't realize. 

The caching implementation and when to use what is years old. I personally would optimize one after the other, e.g. disable the cache entirely and see if things even work better for your use cases.

Another factor, which in many cases makes huge difference a is the readchunk. While it heavily sucks for BluRay isos to have large chunks, as it kills network performance on low CPU devices like raspberry pi2, it enhances throughout a lot for others.

Back when it was implemented, there were still many single core machines and they could either read via network or play a file :-). Others tried sharing there local content via the internet over semi capable lines with additional caches in between.

So in short: no setting will help here for all usecases. I would start with the VFS layer to see e.g. the chunks, seeks, etc. to find out the performance you get per file. Leave out the cache in between completely to find out what it should improve, fix depending on what is not working depending on your tests.

**Comment by fritsch** [comment]
> In this specific case we're talking about the [Mecool KM7](https://www.mecool.com/products/2021-mecool-km7-android-tv-box-android-11-prime-video-youtube-supported-4gb-ram-64gb-rom) with a rather powerful [Amlogic S905y4](https://androidpctv.com/amlogic-s905y4-comparative/)
> 
> There is no SD card or HDD interface in this mix. It has eMMC. Now it's till android with selectively optimized features (like hardware video decoding is awesome). I was wrong about network throughput though. I thought it had gigabit but it has 10/100 mbit. I do get high up in those speeds (like ~80 mbit/sec) when testing it a file transfer in another app.
> 
> We're still talking about SMB here in kodi, not local on-device storage.

Samba with libsmb is working but to my knowledge was never optimized, especially not for throughout. I would suggest to start as outlined above. Chunks, seeks, delays, number of reads, etc.

**Comment by arnova** [comment]
@fritsch: I've said it many times before: Many things don't work properly without filesystem cache. And that's simply because VideoPlayer cache caches different things from the filesystem cache (and the other way around). Also filesystem cache uses 2 buffers to workaround issues with files with bad-interleaving...

I think if the cache setting were to be exposed in the GUI, it should be on a per-source basis. Else it would still be a half-baked solution imo.....

**Comment by markg85** [comment]
> I think if the cache setting were to be exposed in the GUI, it should be on a per-source basis. Else it would still be a half-baked solution imo.....

That would make the feature DOA (dead on arrival).
I would not want to change the cache setting on a per-source basis. Per source basis things should be reserved for source specifics (like audio stream to choose, subtitle, ...)

If you mean per "transport" basis then it would make more sense. So to specify a cache size on SMB and a different one one HTTP, etc...

**Comment by Dogway** [comment]
@markg85 The current defaults are:

```
<advancedsettings>
  <network>
    <buffermode> 1 </buffermode>
    <readbufferfactor> 4 </readbufferfactor>
    <memorysize> 0 </memorysize>
  </network>
</advancedsettings>
```

**Comment by basilgello** [comment]
By the way, do you use embedded smb handler on Linux, vfs.smb2 or mount smb via fstab on your working Linux box and on odroids?

**Comment by markg85** [comment]
@Dogway 

No `memorysize`?
That doesn't match the code.... 
I get nothing of this wealth of inconsistency... what is it **actually** using by default if it's not the code i found??

In your claimed defaults it uses the local storage (eMMC in this case) as cache.
Should be plenty fast.. but still...

Things don't add up here. Those claimed actual defaults should be fine. Yet buffering is a thing.
The settings i use right now:
```
<advancedsettings>
  <network>
    <buffermode>1</buffermode>
    <readbufferfactor>2</readbufferfactor>
    <cachemembuffersize>100000000</cachemembuffersize>
  </network>
</advancedsettings>
```

And my buffering is gone for files served over SMB. Just tried with one movie of ~20GB that used to have buffering every few seconds...

**Comment by markg85** [comment]
> By the way, do you use embedded smb handler on Linux, vfs.smb2 or mount smb via fstab on your working Linux box and on odroids?

On my linux desktop where i've installed kodi just using pacman (archlinux). There i just use SMD resources exactly the same way as i would on the embedded devices. Add a video folder in kodi in it's UI with it's SMB stuff. There is no local fstab or mounting going on.

**Comment by fritsch** [comment]
> @fritsch: I've said it many times before: Many things don't work properly without filesystem cache. And that's simply because VideoPlayer cache caches different things from the filesystem cache (and the other way around). Also filesystem cache uses 2 buffers to workaround issues with files with bad-interleaving...
> 
> I think if the cache setting were to be exposed in the GUI, it should be on a per-source basis. Else it would still be a half-baked solution imo.....

Yeah - and for many years we have complains about kodi network lags / buffers while OS mounted shares are fine at least from some people, while the majority seems to have a good working setup. So, it seems no matter what was / is said - it seems not to work properly for some people, right?

That's why I kindly asked the people with issues to not iterate over the far too complicated and too many advancedsetings, but to start from scratch to understand the system to find out what works and where the culprits are. I did not blame anyone / anything in my post - I only asked to start from the VFS calls, bypassing the cache, trying with just videoplayer and improve / continue from there. Fixing a complicated system for all use-cases won't be successful, it will only add mitigations after mitigations and special cases on top. So as said - for finding bottlenecks, the "direct path" is the most easiest to debug ... 

With plenty of different "files" out there and especially the bluary /dvd rips not being great for streaming by design will for sure need other behaviour than mp4 files read from the start towards ending.

**Comment by basilgello** [comment]
Let's discuss it here: https://forum.kodi.tv/showthread.php?tid=370978

**Comment by arnova** [comment]
> > I think if the cache setting were to be exposed in the GUI, it should be on a per-source basis. Else it would still be a half-baked solution imo.....
> 
> That would make the feature DOA (dead on arrival). I would not want to change the cache setting on a per-source basis. Per source basis things should be reserved for source specifics (like audio stream to choose, subtitle, ...)
> 
> If you mean per "transport" basis then it would make more sense. So to specify a cache size on SMB and a different one one HTTP, etc...

Why would this feature make it DOA?  Perhaps you misunderstood: I meant having an option per source to enable/disable cache, the rest of the settings should still be gloval. Especially if you implement it the way I suggest it (on a per source basis), it will give the user maximum flexibility....
